File: credit.c
#include<stdio.h>
#include<cs50.h>
#include<math.h>

int GetKol (long long number);
int GetLegit(long long number0, int kol);
int GetNameCard(long long number, int kol);

int main(void)
{
    long long number;
    do
    {
        printf("Number: ");
        number = GetLongLong();
    }
    while (number < 0);
    
    long long number0 = number;
    
    int kol = GetKol(number);
    if (kol == 13 || kol == 15 || kol == 16) //shchob marno ne schytaty
    {
        int x = GetLegit(number0, kol);
    
        if (x % 10 == 0)
            GetNameCard(number, kol);
        else 
            printf("INVALID\n");
    }        
    else 
            printf("INVALID\n");
}


int GetKol (long long number)
{
    int kol = 0;        //strlen(number);
    long number_s = number;
    
    do
    {
        kol++;
        number_s = number_s / 10;
    } 
    while (number_s  > 0);
    
    return kol;
}

int GetLegit(long long number0, int kol)
{
    int ch = 0; int ch_double =0; 
    int suma = 0, ch_suma1 = 0, ch_suma2 = 0;
    
    int j = 1 ;
   
    for (int i = 0; i < kol; i++)
    {
        if (j == 2)
        {
            ch = number0 % 10; //ostatok
            ch_double = ch*2;   
            
            if(ch_double > 9) 
                ch_double = (ch_double - 10) + (1);
                
            ch_suma1 = ch_suma1 + ch_double;
            j = 1;
            ///printf("ch= %i\n", ch);printf("ch_double= %i\n", ch_double);printf("ch_suma1= %i\n", ch_suma1);
        }
        else
        {
            ch = number0 % 10;
            ch_suma2 = ch_suma2 + ch;
            ///printf("ch= %i\n", ch);printf("ch_suma2= %i\n", ch_suma2);            
            j = 2;
        }
        number0 = (number0 / 10);  
        ///printf("number0= %lli\n\n", number0);
    }
    suma = ch_suma1 + ch_suma2;
    ///printf("suma= ch_suma1 + ch_suma2= %i\n\n", suma);
    //printf("suma= %i/n", suma);
    return suma;
       
       
}

int GetNameCard(long long number, int kol)
{
    switch(kol)
    {
        case 15:    //start with 34, 37 Amex
            number = number / 10000000000000;
            if((number == 34) || (number == 37)) printf("AMEX\n"); else printf("INVALID\n");
        break;
        
        case 13:    //13 i 16 if... //start with 4 Visa
            number = number / 1000000000000;
            if(number == 4)  printf("VISA\n"); else printf("INVALID\n");
        break;
        
        case 16:    //start with 51 - 55 MasterCard
            number = number / 100000000000000;
            if(number / 10 == 4) printf("VISA\n"); 
                else if((number >= 51) || (number <= 55)) printf("MASTERCARD\n"); 
                    else printf("INVALID\n");
        break;
        
        default: printf("INVALID\n");
        break;
    }
    
    number = number * 1;
    kol = kol * 1;
    //printf("\nYour number: %lli legit, kol = %i\n", number, kol);
   
    return 1;
    
}

File: credit.py
import cs50
import sys

def main():
    while True:
        print("Number: ", end="")
        number = cs50.get_int()
        if not (number < 0):
            break
    # print("number = {}".format(number))
    number0 = number
    kol = GetKol(number)
    # print("kol = {}".format(kol))
    if ((kol == 13) or (kol == 15) or (kol == 16)):
        x = GetLegit(number0, kol)
        if (x % 10 == 0):
            GetNameCard(number, kol)
        else: 
            print("INVALID")
    else:
        print("INVALID")

def GetKol(number):
    #print("number in GetKol= {}".format(number))
    kol = len(str(number))
    #print("kol for GetKol = {}".format(kol))
    return kol

def GetLegit(number0, kol):
    ch = 0
    ch_double = 0 
    suma = 0
    ch_suma1 = 0
    ch_suma2 = 0
    j = 1
    for i in range(kol):
        if j == 2:
            ch = number0 % 10 # ostatok
            ch_double = ch*2  
            if ch_double > 9: 
                ch_double = (ch_double - 10) + (1)
            ch_suma1 = ch_suma1 + ch_double
            j = 1
        else:
            ch = number0 % 10
            ch_suma2 = ch_suma2 + ch
            j = 2
        number0 = (number0 // 10)
    suma = ch_suma1 + ch_suma2
    # print("suma = {}".format(suma))
    return suma

def GetNameCard(number, kol):
    if kol == 15:
        number = number // 10000000000000
        if((number == 34) or (number == 37)): 
            print("AMEX") 
        else: 
            print("INVALID")
    if kol == 13:
        number = number // 1000000000000
        if(number == 4):  
            print("VISA") 
        else:
            print("INVALID")
    if kol == 16:    
        number = number // 100000000000000
        if (number // 10 == 4): 
            print("VISA")
        elif ((number >= 51) or (number <= 55)):
            print("MASTERCARD") 
        else:
            print("INVALID")
    if not (kol == 13 or kol == 15 or kol == 14):
        print("INVALID")
    number = number * 1
    kol = kol * 1
    return 1

if __name__ == "__main__":
    main()